# Exploratory Data Analysis - Instacart Grocery Basket

## Introduction

This analysis explores the Instacart Online Grocery Basket dataset, which contains over 3 million grocery orders from 200,000 users. The dataset includes detailed transaction data with product information and shopping frequencies that we're going to use forbasket analysis and customer behavior insights.

**Objectives :**
- Understand product catalog structure and sales distribution
- Analyse customer ordering patterns and behaviors
- Identify basket composition characteristics
- Generate insights for association rules mining and recommendations

**Structure :**
- I. Data overview and quality
- II. Product catalog analysis
- III. Customer behavior analysis
- IV. Basket composition analysis
- V. Key findings and next steps


In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt

## I. Data overview and quality

### I.1 Dataset loading

In [66]:
# Here we import the function to load the data and the functions to create the visualizations
import sys

sys.path.append('../src/')

import importlib

import visualizations
import load_data

importlib.reload(visualizations) # to force the reloading of the visualizations module
importlib.reload(load_data) # to force the reloading of the load_data module

from load_data import load_instacart_data, quick_info, create_products_enriched

from visualizations import  (
    plot_distribution_of_orders,
    plot_order_hours, 
    plot_orders_by_dow, 
    plot_aisles_per_department, 
    plot_sales_by_department_barplot, 
    plot_sales_by_department_boxplot, 
    plot_sales_by_aisle_barplot, 
    plot_sales_by_aisle_boxplot, 
    plot_top_20_products_barplot, 
    scatterplot_top_100_products,
    plot_basket_size_distribution,
    plot_departments_per_basket,
    plot_basket_size_by_segment,
    plot_basket_size_vs_nb_orders,
    plot_days_since_prior_order_distribution,
    plot_department_aisle_treemap
)


In [17]:
data = load_instacart_data()

orders = data['orders']
order_products_prior = data['order_products_prior']
order_products_train = data['order_products_train']
products = data['products']
aisles = data['aisles']
departments = data['departments']


### I.2 Dataset overview

First we are going to look at the length of the different dataset to understand how many customers, products and transactions we have.

In [12]:
print("="*60)
print("INSTACART DATASET OVERVIEW")
print("="*60)

# Customers
nb_users = orders['user_id'].nunique()
print(f"\n 👤 CUSTOMERS:")
print(f"  - Total unique customers: {nb_users:,}")

# Orders
nb_orders = len(orders)
avg_orders_per_user = orders.groupby('user_id')['order_id'].count().mean()
print(f"\n🛒 ORDERS:")
print(f"  - Total orders: {nb_orders:,}")
print(f"  - Average orders per customer: {avg_orders_per_user:.1f}")

# Products
print(f"\n📦 PRODUCTS:")
print(f"  - Total products: {len(products):,}")
print(f"  - Departments: {len(departments)}")
print(f"  - Aisles: {len(aisles)}")

# Transactions
nb_transactions = len(order_products_prior) + len(order_products_train)
print(f"\n📊 TRANSACTIONS:")
print(f"  - Total product purchases: {nb_transactions:,}")
print(f"  - Prior set: {len(order_products_prior):,}")
print(f"  - Train set: {len(order_products_train):,}")

INSTACART DATASET OVERVIEW

 👤 CUSTOMERS:
  - Total unique customers: 206,209

🛒 ORDERS:
  - Total orders: 3,421,083
  - Average orders per customer: 16.6

📦 PRODUCTS:
  - Total products: 49,688
  - Departments: 21
  - Aisles: 134

📊 TRANSACTIONS:
  - Total product purchases: 33,819,106
  - Prior set: 32,434,489
  - Train set: 1,384,617


### I.3 Data quality

In [ ]:
quick_info(orders, "Orders")
quick_info(order_products_prior, "Order Products Prior")
quick_info(order_products_train, "Order Products Train")
quick_info(products, "Products")
quick_info(departments, "Departments")
quick_info(aisles, "Aisles")

**Key findings :**
- ✅ Data quality is good with only one feature with missing values
- ⚠️ 206,209 missing values in 'days_since_prior_order' -> these are first orders (same number as clients)
- ✅ No duplicates detected
- 📊 Dataset is large especially order_products_prior with 32M rows

## II. Product catalog analysis

### II.1 Catalog structure

For understanding more the product catalog structure, we are going to merge different datasets. I merge it on order_products_prior which is grouped by product_id to get the frequency of purchase for each products. Here we just call the function we've made to merge the datasets.

In [21]:
# compute product sales with information from products, aisles and departments

products_enriched= create_products_enriched(order_products_prior, products, aisles, departments)

# export the enriched products dataset to a csv file for memory efficiency in the next steps of the analysis
products_enriched.to_csv('../data/processed/products_enriched.csv', index=False)

In [22]:
print("="*60)
print("PRODUCT CATALOG STRUCTURE")
print("="*60)

# Products with sales
products_with_sales = len(products_enriched)
products_without_sales = len(products) - products_with_sales

print(f"\n📦 PRODUCTS:")
print(f"  - With sales: {products_with_sales:,} ({products_with_sales/len(products)*100:.1f}%)")
print(f"  - Without sales: {products_without_sales} ({products_without_sales/len(products)*100:.2f}%)")

print(f"\n🏢 HIERARCHY:")
print(f"  - Departments: {products_enriched['department'].nunique()}")
print(f"  - Aisles: {products_enriched['aisle'].nunique()}")
print(f"  - Avg products per department: {len(products_enriched) / products_enriched['department'].nunique():.0f}")
print(f"  - Avg products per aisle: {len(products_enriched) / products_enriched['aisle'].nunique():.0f}")

print(f"\n📋 DEPARTMENTS LIST:")
for i, dept in enumerate(sorted(products_enriched['department'].unique()), 1):
    nb_products = len(products_enriched[products_enriched['department'] == dept])
    print(f"  {i:2d}. {dept:25s} ({nb_products:,} products)")

PRODUCT CATALOG STRUCTURE

📦 PRODUCTS:
  - With sales: 49,677 (100.0%)
  - Without sales: 11 (0.02%)

🏢 HIERARCHY:
  - Departments: 21
  - Aisles: 134
  - Avg products per department: 2366
  - Avg products per aisle: 371

📋 DEPARTMENTS LIST:
   1. alcohol                   (1,054 products)
   2. babies                    (1,081 products)
   3. bakery                    (1,516 products)
   4. beverages                 (4,363 products)
   5. breakfast                 (1,114 products)
   6. bulk                      (38 products)
   7. canned goods              (2,092 products)
   8. dairy eggs                (3,448 products)
   9. deli                      (1,322 products)
  10. dry goods pasta           (1,858 products)
  11. frozen                    (4,007 products)
  12. household                 (3,083 products)
  13. international             (1,139 products)
  14. meat seafood              (907 products)
  15. missing                   (1,255 products)
  16. other                 

There are 49 677 products with sales.

The products are organised into 21 departments and 134 aisles.

We also see there is a department called 'missing' and the top 1 aisle in term of number products is also called 'missing'. We can assume it is for products that are not categorized. Let's check if the department 'missing' contains only the aisle 'missing'. Let's do the same for the category 'other'.



In [25]:
# Special categories : 'missing' vs 'other'


special_categories = ['missing', 'other']
total_sales = products_enriched['nb_sales'].sum()

for cat in special_categories:
    spe_cat= products_enriched[products_enriched['department'] == cat]

    print(f"\n Department: '{cat}'")
    print(f"  Products: {len(spe_cat):,} ({len(spe_cat)/len(products_enriched)*100:.2f}% of catalog)")
    print(f"  Total sales: {spe_cat['nb_sales'].sum():,} ({spe_cat['nb_sales'].sum()/total_sales*100:.2f}% of sales)")
    print(f"  Unique aisles: {spe_cat['aisle'].nunique()}")
    
    if spe_cat['aisle'].nunique() == 1:
        print(f"  Aisles: {list(spe_cat['aisle'].unique())}")
    else:
        print(f"  Aisles: Multiple aisles present")

    print(f"\n  Top 5 products:")
    top_5_names = spe_cat.nlargest(5, 'nb_sales')['product_name'].tolist()
    for name in top_5_names:
        print(f"    - {name}")




 Department: 'missing'
  Products: 1,255 (2.53% of catalog)
  Total sales: 69,145 (0.21% of sales)
  Unique aisles: 1
  Aisles: ['missing']

  Top 5 products:
    - Organic Riced Cauliflower
    - Peanut Butter Ice Cream Cup
    - Organic Mango Yogurt
    - Organic Celery Bunch
    - Natural Lemon Flavored Sparkling Water

 Department: 'other'
  Products: 548 (1.10% of catalog)
  Total sales: 36,291 (0.11% of sales)
  Unique aisles: 1
  Aisles: ['other']

  Top 5 products:
    - Roasted Almond Butter
    - Light CocoWhip! Coconut Whipped Topping
    - 93/7 Ground Beef
    - Coconut Almond Granola
    - Coffee Mate French Vanilla Creamer Packets


This confirms the hypothesis, the departments 'missing' and 'other' contains only 1 aisle with the same name. When we look at the top 5 products of each category it seems they don't have any link. 
Those products represent less than 1% of the sales so we're not going to dig more into it.

### II.2 Sales distribution by departments and aisles

In [12]:
aisles_per_department = products_enriched.groupby('department')['aisle'].nunique().reset_index(name='nb_aisles').sort_values('nb_aisles', ascending=False)

fig = plot_aisles_per_department(aisles_per_department)
fig

/tmp/ipykernel_23192/1601575320.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  aisles_per_department = products_enriched.groupby('department')['aisle'].nunique().reset_index(name='nb_aisles').sort_values('nb_aisles', ascending=False)


We see that all departments contain in between 1 and 17 aisles. The department 'personal care' is the one with the higher count of aisles this mean there is a lot of diversity inside this department.

In [51]:
fig = plot_department_aisle_treemap(products_enriched, top_n_per_dept=4)
fig.show()

The department 'produce' (fruits and vegetables) represent the highest number of sales (29% of the total) 

The second one is 'dairy eggs' (milk, cheese, ...) with more than 16%. 

We also see that the deppartments with the highest number of sales are not the ones with the higher number of products.

In [67]:
# Let's visualize the distribution of product sales per department
fig=plot_sales_by_department_boxplot(products_enriched)
fig

/home/francesco21/Documents/Retail-Data-Insights/notebooks/../src/visualizations.py:100: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



We can see there are a high variability of sales inside each department with a lot of outliers. This means that some products have a big impact on sales and it's probably on them that we'll focus.

In [68]:
top_aisles = products_enriched.groupby('aisle')['nb_sales'].sum().nlargest(20).index
sales_by_aisle = products_enriched[products_enriched['aisle'].isin(top_aisles)]

/tmp/ipykernel_8551/2551244913.py:1: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [69]:
# Let's visualize the distribution of product sales per aisle for the top 20 aisles
fig=plot_sales_by_aisle_boxplot(sales_by_aisle)
fig

/home/francesco21/Documents/Retail-Data-Insights/notebooks/../src/visualizations.py:139: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



We also find a high variability of sales within aisles. For example, the top sales of "fresh fruits" are much higher than the sales of "fresh vegetables".

Let's check the importance of those 20 aisles on the total sales :

In [64]:
total_sales = products_enriched['nb_sales'].sum()
total_products = len(products_enriched)

top_20_sales = top_aisles_sales['total_sales'].sum()
top_20_products = top_aisles_sales['nb_products'].sum()

percentage_sales = (top_20_sales / total_sales) * 100
percentage_products = (top_20_products / total_products) * 100

print(f"  - {top_20_products:,} products ({percentage_products:.1f}% of catalog)")
print(f"  - {top_20_sales:,} sales ({percentage_sales:.1f}% of total)")


  - 11,794 products (23.7% of catalog)
  - 19,445,677 sales (60.0% of total)


As we can see those 20 aisles represent 60% of the total sales. But we need to pay attention that we only focus here on the quantity of sales and not the value of sales. We can assume that some products such as 'meat' or 'personal care' are more expensive and can have a big impact on the results of the company.

### Focus on products

In [ ]:
top_20_products = products_enriched.nlargest(20, 'nb_sales')

fig=plot_top_20_products_barplot(top_20_products)
fig

The top 20 products in term of sales are represented by 19 products from 'Produce' department and 1 from 'Dairy eggs'. The top products in vegetables and fruits are from far the highest numbers of sales in Instacart !

It would be interesting to check quickly the relation in between high sales and reorder rate.

In [ ]:
# Let's have a look at the reorder rate vs number of sales for the top 100 products

top_100_products = products_enriched.nlargest(100, 'nb_sales')

fig=scatterplot_top_100_products(top_100_products)
fig

We can see for example for bananas that they have also a high reorder rate. When we are looking at the other products we see that a high number of sales is not always linked to a high reorder rate. 

The key takeaways of the analysis of product sales are :

Hierarchy : 21 departments -> 134 aisles -> 49677 products

Dominant categories : 

    - Departments : Produce (29%), Dairy eggs (17%), Snacks (9%)

    - Aisles : Fresh fruits and vegetables, packaged vegetables and yogurt

    - Products : Bananas, organic bananas and organic strawberries are the best sellers

Strategic insights and limitation :

    - We saw that the top 20 aisles represent 60% of the total sales

    - We need to pay attention that we only focus on the frequency of sales. We don't have the quantity of each products in one order

    - It would be interesting to add prices to not simply compare the frequency of sales but also the values


## 2. Orders dataset analysis

bla bla bla

In [21]:
quick_info(orders, "Orders")


Orders: 3,421,083 rows × 7 columns
Missing: 206209
Duplicates: 0
order_id                     int32
user_id                      int32
eval_set                  category
order_number                  int8
order_dow                     int8
order_hour_of_day             int8
days_since_prior_order     float32
dtype: object


Around 200k lines on 3.4M  with missing values on 'days_since_prior_order' => assuming it means it's the first order of the user ?

In [ ]:
user_order_counts = orders.user_id.value_counts().reset_index()
user_order_counts.columns = ['user_id', 'nb_orders']

user_order_counts['nb_orders'].describe()

We find again 206 209 different user_id so the hypothesis of value missing on 'days_since_prior_order' is correct.
We also see we have in between 4 and 100 commands per client in the database

In [ ]:
fig=plot_distribution_of_orders(user_order_counts)
fig 

The average is way higher than then median. We have a right skewed distribution.
50% of the clients made less than 10 orders and 25% of clients made between 20 and 100 orders. 
It would be interesting to see the impact of the extreme values meaning important customers.

In [ ]:
# Segments
occasional = user_order_counts[user_order_counts['nb_orders'] <= 10]
regular = user_order_counts[(user_order_counts['nb_orders'] > 10) & (user_order_counts['nb_orders'] <= 30)]
heavy = user_order_counts[user_order_counts['nb_orders'] > 30]

print(f"USER SEGMENTATION BASED ON ORDER FREQUENCY")
print(f"\n Casual customers (≤10 orders):")
print(f"   - Number : {len(occasional):,} users ({len(occasional)/len(user_order_counts)*100:.1f}%)")
print(f"   - Mean : {occasional['nb_orders'].mean():.1f} orders")

print(f"\n Regular customers (10-30 orders):")
print(f"   - Number : {len(regular):,} users ({len(regular)/len(user_order_counts)*100:.1f}%)")
print(f"   - Mean : {regular['nb_orders'].mean():.1f} orders")

print(f"\n Heavy customers (>30 orders):")
print(f"   - Number : {len(heavy):,} users ({len(heavy)/len(user_order_counts)*100:.1f}%)")
print(f"   - Mean : {heavy['nb_orders'].mean():.1f} orders")

# Impact on total orders
total_orders = user_order_counts['nb_orders'].sum()
heavy_orders = heavy['nb_orders'].sum()

print(f"\n IMPACT BUSINESS:")
print(f"   - Heavy users = {len(heavy)/len(user_order_counts)*100:.1f}% of users")
print(f"   - Generate {heavy_orders/total_orders*100:.1f}% of orders")

In [ ]:
user_order_counts['segment'] = 'Casual'
user_order_counts.loc[user_order_counts['nb_orders'] > 10, 'segment'] = 'Regular'
user_order_counts.loc[user_order_counts['nb_orders'] > 30, 'segment'] = 'Heavy'
user_order_counts.head()    

Let's have a look to the distribution of the orders by day. We'll use the US convention meaning that 0 is on Sunday.

In [ ]:
# Let's have a look to the distribution of orders by day of the week
fig = plot_orders_by_dow(orders)
fig

As we can see, there are more orders made on Sunday and Monday and less in the middle of the week.

In [ ]:
# Let's have a look to the distribution of orders by hour of the day
fig=plot_order_hours(orders)
fig

In [ ]:
# we want to analyse the frequency of orders
orders_with_days= orders.dropna(subset=['days_since_prior_order'])

fig = plot_days_since_prior_order_distribution(orders_with_days)
fig 

In [ ]:
orders_with_days['days_since_prior_order'].describe()

The average of days in between two orders is around 11 days. 

It's qutie surprising the high quantity of orders with a value of days_since_prior_order of 30. We can assume there is a maximum value for this field. 

## Analyse of basket size

In [ ]:
order_products_prior.head()

In [ ]:
order_products_prior.reordered.value_counts()


In [ ]:
order_products_enriched = order_products_prior[['order_id', 'product_id']].merge(products_enriched[['product_id', 'product_name', 'department', 'aisle']],on='product_id', how='inner')

In [ ]:
basket_size = (
    order_products_enriched
    .groupby('order_id')
    .size()
    .reset_index(name='basket_size')
)

print("BASKET SIZE STATISTICS")
print(f"  Mean   : {basket_size['basket_size'].mean():.2f}")
print(f"  Median : {basket_size['basket_size'].median():.0f}")
print(f"  P90    : {basket_size['basket_size'].quantile(0.90):.0f}")
print(f"  Min    : {basket_size['basket_size'].min()}")
print(f"  Max    : {basket_size['basket_size'].max()}")

fig = plot_basket_size_distribution(basket_size)
fig

We have a right-skewed distribution. The majority of the baskets are small with an average of 10 products per order.

We can see that 90 percent of the orders contain less than 20 products so we'll probably not work much on the big orders which are more rare.

In [ ]:
dept_per_basket = (order_products_enriched.groupby('order_id')['department'].nunique().reset_index(name='nb_departments'))

print("DEPARTMENTS PER BASKET")
print(f"  Mean   : {dept_per_basket['nb_departments'].mean():.2f}")
print(f"  Median : {dept_per_basket['nb_departments'].median():.0f}")
print(f"  P90    : {dept_per_basket['nb_departments'].quantile(0.90):.0f}")
print(f"  Min    : {dept_per_basket['nb_departments'].min()}")
print(f"  Max    : {dept_per_basket['nb_departments'].max()}")

fig = plot_departments_per_basket(dept_per_basket)
fig

In average the orders contain products from less than 5 departments. We also have extreme values with a maximum of 18 departments on 21 !

90% of the orders contain products from 1 to 8 departments.



It would be interesting to look at the size of the basket for the 3 types of customers in term of number of orders (occasional, regular and heavy)

In [20]:
basket_with_user = basket_size.merge(orders[['order_id', 'user_id']], on='order_id', how='left')
basket_by_segment = basket_with_user.merge(user_order_counts[['user_id', 'segment']], on='user_id', how='left')

fig = plot_basket_size_by_segment(basket_by_segment)
fig

NameError: name 'basket_size' is not defined

We can see that we can find small basket sizes for all types of customers. But we can also observe that the highest average basket sizes are not for the heaviest customers in term of number of orders. This could be explained by the fact that heavy users tend to order more frequently but with smaller baskets. 

It seems also that sometimes occasionnal customers can have quite large basket sizes, which could indicate that they place large orders infrequently. 

When we look at the previous boxplot we can still see that the largest baskets are not ordered by the occasional customers.

## Focus on last item purchased

We would like to see if we have recurrent items bought in last in each orders to see if there are some compulsive purchases at the end of a basket.

In [11]:
# we want to keep only the last item for each basket in the 'order_products_prior' dataset
# we need to merge it with products to have the product name

last_order = (
    order_products_prior
    .sort_values(['order_id', 'add_to_cart_order'])  
    .drop_duplicates(subset='order_id', keep='last')  
    .merge(products[['product_id', 'product_name']], on='product_id', how='inner')
)
last_product_freq = (
    last_order['product_name']
    .value_counts()
    .reset_index()
    .rename(columns={'index': 'product_name', 'product_name': 'last_position_count'})
)

print("Top 20 produits en dernière position :")
print(last_product_freq.head(20))

Top 20 produits en dernière position :
       last_position_count  count
0                   Banana  30098
1   Bag of Organic Bananas  29826
2     Organic Strawberries  23477
3     Organic Baby Spinach  21733
4     Organic Hass Avocado  15544
5             Strawberries  15529
6          Organic Avocado  12546
7              Large Lemon  11883
8      Organic Blueberries  11347
9       Organic Whole Milk  11003
10                   Limes  10405
11     Organic Raspberries   9782
12            Spring Water   8715
13        Organic Zucchini   8375
14                    Soda   8363
15     Seedless Red Grapes   7795
16    Organic Yellow Onion   7777
17     Organic Half & Half   7739
18          Organic Garlic   7538
19      Organic Fuji Apple   7481


In [14]:
import pandas as pd

# Garde le dernier produit de chaque commande
last_order = (
    order_products_prior
    .sort_values(['order_id', 'add_to_cart_order'])
    .drop_duplicates(subset='order_id', keep='last')
    .merge(products[['product_id', 'product_name']], on='product_id')
)

# Fréquence en dernier
last_product_freq = last_order['product_name'].value_counts().reset_index()
last_product_freq.columns = ['product_name', 'last_position_count']

# Fréquence générale
general_freq = (
    order_products_prior
    .merge(products[['product_id', 'product_name']], on='product_id')
    ['product_name']
    .value_counts()
    .reset_index()
)
general_freq.columns = ['product_name', 'total_count']

# Merge
comparison = last_product_freq.merge(general_freq, on='product_name')

# Calcule les ratios
total_orders = order_products_prior['order_id'].nunique()
comparison['last_position_rate'] = comparison['last_position_count'] / total_orders
comparison['general_rate'] = comparison['total_count'] / len(order_products_prior)
comparison['impulse_ratio'] = comparison['last_position_rate'] / comparison['general_rate']

# Trie
impulse_products = comparison.sort_values('impulse_ratio', ascending=False)

# ===== FILTRE PAR VOLUME (CRITIQUE !) =====
impulse_significant = impulse_products[
    (impulse_products['last_position_count'] >= 2000) &   # 2000 minimum
    (impulse_products['total_count'] >= 10000) &          # 10000 minimum
    (impulse_products['impulse_ratio'] > 1.5)             # 30% plus fréquent
]

print(f"\n📊 Filtrage :")
print(f"   Avant filtre : {len(impulse_products)} produits")
print(f"   Après filtre : {len(impulse_significant)} produits")

# Ajoute département
impulse_significant = impulse_significant.merge(
    products[['product_name', 'department_id']].drop_duplicates(),
    on='product_name',
    how='left'
)

print("\n🎯 TOP 30 IMPULSE PRODUCTS (SIGNIFICATIFS) :")
print(impulse_significant.head(30)[['product_name', 'department_id', 'impulse_ratio', 'last_position_count', 'total_count']])

# Sauvegarde
impulse_significant.to_csv('../data/processed/impulse_products.csv', index=False)
print(f"\n✅ {len(impulse_significant)} produits impulse sauvegardés")


📊 Filtrage :
   Avant filtre : 49688 produits
   Après filtre : 21 produits

🎯 TOP 30 IMPULSE PRODUCTS (SIGNIFICATIFS) :
                      product_name  department_id  impulse_ratio  \
0                  Sparkling Water              7       2.509734   
1                        Trail Mix             19       2.383444   
2                             Soda              7       2.357390   
3         0% Greek Strained Yogurt             16       2.223853   
4                       Smartwater              7       2.218783   
5           Chocolate Chip Cookies             19       2.125739   
6                      Clementines              4       2.111132   
7             Sweet Kale Salad Mix              4       2.100889   
8                           Apples              4       2.074740   
9   Organic Extra Virgin Olive Oil             13       1.979455   
10          Organic Tortilla Chips             19       1.953183   
11         Sparkling Mineral Water              7       1.7754

In [12]:
# Ajoute cette cellule après ton code

# Stats par département
impulse_by_dept = (
    impulse_significant
    .groupby('department_id')
    .agg({
        'product_name': 'count',
        'impulse_ratio': ['mean', 'max'],
        'last_position_count': 'sum',
        'total_count': 'sum'
    })
    .round(2)
)

impulse_by_dept.columns = ['nb_products', 'avg_ratio', 'max_ratio', 'sum_last_pos', 'sum_total']
impulse_by_dept = impulse_by_dept.sort_values('nb_products', ascending=False)

print("\n📦 IMPULSE PRODUCTS PAR DÉPARTEMENT :")
print(impulse_by_dept)

# Merge avec noms de départements si tu as le fichier
try:
    departments = pd.read_csv('../data/raw/departments.csv')
    impulse_by_dept = impulse_by_dept.reset_index().merge(
        departments, 
        on='department_id'
    )
    print("\n📦 AVEC NOMS :")
    print(impulse_by_dept[['department', 'nb_products', 'avg_ratio', 'max_ratio']])
except:
    print("(departments.csv non trouvé)")


📦 IMPULSE PRODUCTS PAR DÉPARTEMENT :
               nb_products  avg_ratio  max_ratio  sum_last_pos  sum_total
department_id                                                            
4                       27       1.63       2.45         64161     421021
7                       24       1.89       3.70         64585     373296
19                      16       1.99       3.04         29571     153022
13                      10       1.37       1.98         23809     178098
16                       9       1.67       2.22         15522      94198
1                        4       1.39       1.74          5997      43371
17                       3       1.39       1.44          7767      56869
5                        2       1.74       1.75          2475      14361
14                       2       1.85       2.15          3472      19636
10                       1       1.49       1.49          1499      10150
15                       1       1.50       1.50          1563      10521
